backend/drm/drm.c

```diff
@@ -1616,3 +1629,118 @@ void destroy_drm_connector(struct wlr_drm_connector *conn) {
 }
+
+int wlr_drm_backend_get_client_fd(struct wlr_backend *backend) {
+	assert(backend && wlr_backend_is_drm(backend));
+
+	struct wlr_drm_backend *drm = get_drm_backend_from_backend(backend);
+	char *path = drmGetDeviceNameFromFd2(drm->fd);
+
+	int fd = open(path, O_RDWR);
+	if (fd < 0) {
+		wlr_log_errno(WLR_ERROR, "Unable to clone DRM fd for client fd.\n");
+		free(path);
+		return -EINVAL;
+	}
+	if (drmIsMaster(fd)) {
+		drmDropMaster(fd);
+	}
+	assert(!drmIsMaster(fd) && "Won't send master fd to client");
+	free(path);
+	return fd;
+}
+
+int wlr_drm_backend_create_lease(struct wlr_backend *backend,
+		struct wlr_output **outputs, size_t n_outputs, uint32_t *lessee_id) {
+	assert(backend && wlr_backend_is_drm(backend));
+	struct wlr_drm_backend *drm = get_drm_backend_from_backend(backend);
+
+	int n_objects = 0;
+	for (size_t i = 0; i < n_outputs; ++i) {
+		struct wlr_output *output = outputs[i];
+		if (!wlr_output_is_drm(output)) {
+			wlr_log(WLR_ERROR, "Output '%s' is not a DRM output", output->name);
+			return -EINVAL;
+		}
+
+		struct wlr_drm_connector *conn = (struct wlr_drm_connector *)output;
+		assert(conn->state != WLR_DRM_CONN_LEASED);
+
+		if (conn->backend != drm) {
+			wlr_log(WLR_ERROR, "Output '%s' doesn't belong to provided backend",
+					output->name);
+			return -EINVAL;
+		}
+
+		/* conn->id + conn->crtc->id + conn->crtc->primary->id */
+		n_objects += 3 + (conn->crtc->cursor ? 1 : 0);
+	}
+
+	if (n_objects == 0) {
+		wlr_log(WLR_ERROR, "Attempting to create a DRM lease with 0 objects");
+		return -EINVAL;
+	}
+
+	uint32_t objects[n_objects + 1];
+	for (size_t i = 0, j = 0; i < n_outputs; ++i) {
+		struct wlr_drm_connector *conn = (struct wlr_drm_connector *)outputs[i];
+		objects[j++] = conn->id;
+		objects[j++] = conn->crtc->id;
+		objects[j++] = conn->crtc->primary->id;
+		wlr_log(WLR_DEBUG, "connector: %d crtc: %d primary plane: %d",
+				conn->id, conn->crtc->id, conn->crtc->primary->id);
+		if (conn->crtc->cursor) {
+			wlr_log(WLR_DEBUG, "cursor plane: %d", conn->crtc->cursor->id);
+			objects[j++] = conn->crtc->cursor->id;
+		}
+	}
+
+	wlr_log(WLR_DEBUG, "Issuing DRM lease with the %d objects:", n_objects);
+	int lease_fd = drmModeCreateLease(drm->fd, objects, n_objects, 0,
+			lessee_id);
+	if (lease_fd < 0) {
+		return lease_fd;
+	}
+
+	wlr_log(WLR_DEBUG, "Issued DRM lease %d", *lessee_id);
+	for (size_t i = 0; i < n_outputs; ++i) {
+		struct wlr_drm_connector *conn = (struct wlr_drm_connector *)outputs[i];
+		conn->lessee_id = *lessee_id;
+		conn->crtc->lessee_id = *lessee_id;
+		conn->state = WLR_DRM_CONN_LEASED;
+		wlr_output_destroy(&conn->output);
+	}
+
+	return lease_fd;
+}
+
+int wlr_drm_backend_terminate_lease(struct wlr_backend *backend,
+		uint32_t lessee_id) {
+	wlr_log(WLR_DEBUG, "Terminating DRM lease %d", lessee_id);
+
+	assert(backend && wlr_backend_is_drm(backend));
+	struct wlr_drm_backend *drm = get_drm_backend_from_backend(backend);
+
+	int r = drmModeRevokeLease(drm->fd, lessee_id);
```

Comment:
Getting `ENOENT` during lease termination.

Reply:
Maybe a driver issue?

Reply:
I think the kernel destroys the lease when the client closes its leased DRM FD (see `drm_master_destroy` which calls `drm_lease_destroy`). The kernel will send a `LEASE=1` uevent to notify us that the lease has been destroyed (see `drm_sysfs_lease_event`). Then the client will destroy the Wayland lease object and/or disconnect.

Reply:
We should be able to make the difference between a hotplug event and a lease event with udev properties here: https://github.com/swaywm/wlroots/blob/b2f6db3533f2aff20986ee33ed6364d8bd4553bb/backend/session/session.c#L183

For hotplug events, `udev_device_get_property_value(dev, "HOTPLUG")` will return "1". For lease events, `udev_device_get_property_value(dev, "LEASE")` will return "1".

Reply:
The first commit of https://github.com/swaywm/wlroots/pull/2959 implements part of this.